src/QirRuntime/lib/QIR/arrays.cpp

```diff
@@ -14,35 +14,41 @@
 
 #include "CoreTypes.hpp"
 
-long QirArray::AddRef()
+int QirArray::AddRef()
 {
-    long rc = ++this->refCount;
-    assert(rc != 1); // not allowed to resurrect!
-    return rc;
+    assert(this->refCount != 0 && "Cannot resurrect released array!");

```

Comment:
Is there a reason for this to be an `assert` rather than a `quantum__rt__fail` like in `RemoveUser`.

Reply:
Yes. The user/access count doesn't affect the object's lifetime. Going below zero would represent a logical error in the generated QIR and we can _always_ catch it. Refcounts are different. If a refcount goes to zero, the object is deleted and the memory is released (and potentially reused). So, unless we are lucky and the ghost object is still in memory, the check would be done on, effectively, a random value -- we cannot base production error handling on that, but the assert is OK as far as documentation of the code goes.

I am considering adding a facility for more deterministic memory debugging (to catch leaks and double-releases reliably), but that will be a separate change.